core/src/main/scala/tamer/kafka/Kafka.scala

```diff
@@ -60,15 +60,19 @@ object Kafka {
 
         def mkRegistry(src: SchemaRegistryClient, topic: String) = (ZLayer.succeed(src) >>> Registry.live) ++ (ZLayer.succeed(topic) >>> Topic.live)
 
-        def mkRecordChunk(kvs: List[(K, V)]) = Chunk.fromIterable(kvs.map { case (k, v) => new ProducerRecord(cfg.sink.topic, k, v) })
         def sink(q: Queue[(K, V)], p: Producer.Service[Registry with Topic, K, V], layer: ULayer[Registry with Topic]) =
           logTask.flatMap { log =>
-            q.takeAll.flatMap {
-              case Nil => log.trace("no data to push").unit
-              case kvs =>
-                p.produceChunkAsync(mkRecordChunk(kvs)).provideSomeLayer[Blocking](layer).retry(tenTimes).flatten.unit <*
-                  log.info(s"pushed ${kvs.size} messages to ${cfg.sink.topic}")
-            }
+            ZStream
+              .fromQueue(q)
+              .map { case (k, v) => new ProducerRecord(cfg.sink.topic, k, v) }
+              .groupedWithin(cfg.bufferSize, 1.second)
```

Comment:
Do we even need to explicitly do this? It shouldn't spike the cpu as it takes one element/chunk off the queue

I was also thinking we could offer a chunk to the queue directly and construct this stream below using fromQueueWithShutdown (likely removing the need to do manual fiddling of the queue).

What about adding a test to this?

Reply:
Are you referring to the whole idea of the PR? The CPU sustained 100% usage was confirmed, it went to 5% after this modification. If you mean the `groupedWithin`, why do you think `fromQueueWithShutdown` is what we're looking for?
If we work natively with chunks we have to create them at some point, that just moves the problem in the modules. Decentralizing optimization logic might not give a big advantage.
I’ll add a test once we decide which direction we want to go.

Reply:
No I'm saying that we shouldn't need to groupByWithin to get the CPU down.

Also, I was proposing to consider we create the ZStream.fromQueueWithShutdown to remove handling it via ZManaged as in [here](https://github.com/laserdisc-io/tamer/blob/0e93d2c33c5a066a3af4023e66730ca423cc6c93/core/src/main/scala/tamer/kafka/Kafka.scala#L57).

The chunk part would mean we use [this](https://github.com/zio/zio/blob/f530d6162589f286e1c12656f793326c8466e56e/streams/shared/src/main/scala/zio/stream/ZStream.scala#L3548) instead, given we push the whole chunk instead of [this](https://github.com/laserdisc-io/tamer/blob/0e93d2c33c5a066a3af4023e66730ca423cc6c93/doobie/src/main/scala/tamer/db/package.scala#L60)

Reply:
ok, got it, will do